feat(hardware): ot3: set per-axis max speeds

sfoster1 opened the pull request:
Currently, we can specify a linear speed that an individual move cannot
exceed. This limit is generally given by higher level code and is more
about scientific concerns (avoiding shaking, for instance) than about
robotics concerns.

However, there are also speed limits that are based on robotics
concerns. Any given axis has a speed it may not exceed due to physical
limits - the speed at which its dynamic friction overcomes its available
torque, the speed at which it's no longer possible to represent that
speed correctly in binary, etc - and that limit, too, must be
represented.

These speeds are properly part of an axis's constraints, so we add them
there.

We can implement the speed limit by making the input speed provisioned
on each Move object the smaller of the linear speed limit and the
projected axis-speed-limit-based value when we make a new Move object at
the beginning of a blend pass.

This happens by decomposing the linear speed through the move's unit
vector and comparing each component to the individual axis limit. If the
limit is lower, we can apply ratio between the axis speed limit and the
axis linear speed component to scale down the given speed.

Then, we can use the per-axis max speed values we've always had (and that have
been silently ignored) to actually provision the system constraints.

Closes RCORE-125

sfoster1 committed:
feat(hardware): limit speeds by axis

Currently, we can specify a linear speed that an individual move cannot
exceed. This limit is generally given by higher level code and is more
about scientific concerns (avoiding shaking, for instance) than about
robotics concerns.

However, there are also speed limits that are based on robotics
concerns. Any given axis has a speed it may not exceed due to physical
limits - the speed at which its dynamic friction overcomes its available
torque, the speed at which it's no longer possible to represent that
speed correctly in binary, etc - and that limit, too, must be
represented.

These speeds are properly part of an axis's constraints, so we add them
there.

We can implement the speed limit by making the input speed provisioned
on each Move object the smaller of the linear speed limit and the
projected axis-speed-limit-based value when we make a new Move object at
the beginning of a blend pass.

This happens by decomposing the linear speed through the move's unit
vector and comparing each component to the individual axis limit. If the
limit is lower, we can apply ratio between the axis speed limit and the
axis linear speed component to scale down the given speed.

sfoster1 committed:
refactor(api): ot3: axis max speeds in constraints

We already had per-axis max speed values in our configs, and we... just
weren't doing anything with them. Now, we pass them into the system
constraints used by the hardware motion planner.

sfoster1 committed:
refactor(api): ot3: limit plunger axis max speed

The plunger axis' maximum representable speed is 47mm/s at its current
steps/mm. Limit it to a nice round number below that.

ahiuchingau commented on hardware/opentrons_hardware/hardware_control/motion_planning/move_utils.py:
```diff
@@ -55,8 +55,36 @@ def get_unit_vector(
+            log.info(
+                f"speed {max_linear_speed} decreased by {axis_ratio} because {axis} speed limit is {axis_speed}"
+            )
+            scale = axis_ratio
```
![alt](https://media0.giphy.com/media/HDNcjt5ELkJSE/giphy.gif)

sfoster1 replied:
We can allow people to set their own.... but we should only allow it if their values are _smaller_

ahiuchingau reviewed (approved):
This is awesome! 

Got a question about this function ot3api.py:
``` 
@ExecutionManagerProvider.wait_for_running
    async def _move(
        self,
        target_position: "OrderedDict[OT3Axis, float]",
        speed: Optional[float] = None,
        home_flagged_axes: bool = True,
        max_speeds: Optional[OT3AxisMap[float]] = None,
        acquire_lock: bool = True,
        check_bounds: MotionChecks = MotionChecks.NONE,
    ) -> None:
        """Worker function to apply robot motion."""
```

Now that we have the system-defined `max_speed` per axis, are we going to keep both `speed` & `max_speed` here for user-defined speeds? If so, should we rename one of these `max_speeds` to make it more clear which one we're talking about?

sfoster1 commented:
Tested on hardware and the pipette doesn't make adorable chirps anymore, but also everything is slower because all those config values weren't being restricted, so it's time to increase them

sfoster1 committed:
fix(api): ot3: config updates

Update the axis speed limits now that they are being respected, and bump
the currents.

sfoster1 committed:
format

sfoster1 committed:
need to update tests also
